ts/nni_manager/training_service/local_v3/local.ts

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { Logger, getLogger } from 'common/log';
+import type { LocalConfig, TrainingServiceConfig } from 'common/experimentConfig';
+import type { EnvironmentInfo, Metric, Parameter, TrainingServiceV3 } from 'common/training_service_v3';
+import { TrialKeeper } from 'common/trial_keeper/keeper';
+
+export class LocalTrainingServiceV3 implements TrainingServiceV3 {
+    private config: LocalConfig;
+    private env: EnvironmentInfo;
+    private log: Logger;
+    private trialKeeper: TrialKeeper;
+
+    constructor(trainingServiceId: string, config: TrainingServiceConfig) {
+        this.log = getLogger(`LocalV3.${trainingServiceId}`);
+        this.log.debug('Training sevice config:', config);
+
+        this.config = config as LocalConfig;
+        this.env = { id: `${trainingServiceId}-env` };
+        this.trialKeeper = new TrialKeeper(this.env.id, 'local', Boolean(config.trialGpuNumber));
+    }
+
+    public async init(): Promise<void> {
+        return;
+    }
+
+    public async start(): Promise<EnvironmentInfo[]> {
+        this.log.info('Start');
+        await this.trialKeeper.start();
+        return [ this.env ];
+    }
+
+    public async stop(): Promise<void> {
+        await this.trialKeeper.shutdown();
+        this.log.info('All trials stopped');
+    }
+
+    /**
+     *  Note:
+     *  The directory is not copied, so changes in code directory will affect new trials.
+     *  This is different from all other training services.
+     **/
+    public async uploadDirectory(directoryName: string, path: string): Promise<void> {
+        this.log.info(`Register directory ${directoryName} = ${path}`);
+        this.trialKeeper.registerDirectory(directoryName, path);
+    }
+
+    public async createTrial(_envId: string, trialCommand: string, directoryName: string, sequenceId?: number):
+            Promise<string | null> {
+
+        const trialId = uuid();
```

Comment:
So, a trial is launched without a parameter, and it will get a parameter when it is launched and it "requests" a parameter.
Don't you think it's a a bit complicated? Sometimes (in amulet I imagine), it would be most convenient to send the parameter at job submit. In this way, I don't have to implement a two-way communication, but only metrics collection.

Reply:
But currently by design a trial can request multiple parameters (maybe not documented). That's why we could drop multi-phase mode.
If a trial can only have one parameter, it's a regression.
If a trial can still request more parameters, then pre-allocating parameter is merely a performance optimize.

Reply:
IIUC, I don't know the trial parameter here when it submits. The parameter is only provided after calling `createTrial`.
This brings a headache to create a "trial placeholder" at the beginning, who watches out for new parameters and does nothing until parameters have been received.
Maybe you believe trial keeper is a good idea to handle all these details, but IMHO, this brings an extra level of uncertainty given the chaotic situations of clusters here at Microsoft. As the cluster executes the trial keeper instead of the bare trial command, I feel rather skeptical whether the trial keeper can work transparently when the environment variables, command syntax, distributed training (e.g., MPI), signals get more complicated.

Therefore, I'm asking here is: whether it's possible to invoke `requestParameter` inside `createTrial`. In that case, the training service itself is the one who is asking for new parameters. It can directly submit a trial **with** a parameter.

Reply:
what is the conclusion here?